File: built-in plugins dsp/mb-mic-preamp.cpp
/**
 * MB Mic Preamp
 * Category : effect
 * Type     : microphone
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Vintage mic preamp coloration
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MIC_PREAMP_H
#define MB_MIC_PREAMP_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMicPreamp : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mic-preamp";
    static constexpr const char* PLUGIN_NAME    = "MB Mic Preamp";
    static constexpr const char* PLUGIN_TYPE    = "microphone";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float gain = 20f;  // range [0, 60]
    float impedance = 0.5f;  // range [0, 1]
    float color = 0.3f;  // range [0, 1]
    };

    MbMicPreamp() = default;
    ~MbMicPreamp() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.gain = std::clamp(params.gain, 0f, 60f);
        params.impedance = std::clamp(params.impedance, 0f, 1f);
        params.color = std::clamp(params.color, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Mic Preamp
        return input;
    }
};

#endif // MB_MIC_PREAMP_H

File: built-in plugins dsp/mb-mic-preamp.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbMicPreampPlugin: PluginDefinition = { id: 'mb-mic-preamp', slug: 'mb-mic-preamp', name: 'MB Mic Preamp', category: 'effect', type: 'microphone', version: '1.0.0', description: 'Vintage mic preamp coloration', author: 'Max Booster', parameters: [{ id: 'gain', name: 'Gain', type: 'float', defaultValue: 20, minValue: 0, maxValue: 60, automatable: true }, { id: 'impedance', name: 'Impedance', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'color', name: 'Color', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { gain: 20, impedance: 0.5, color: 0.3 } };

export default MbMicPreampPlugin;
